keps/2937-resource-transformer/README.md

```diff
@@ -0,0 +1,560 @@
+
+## Design Details
+
+Since Kueue's workload.Info struct is already a well-defined
+place where resource filtering is being applied to compute the effective resource
+request from a Workload's PodSpec, the core of the implementation is
+straightforward and follows from how `excludedResourcePrefixes` are already processed.
+
+The non-trivial design issues are how to configure the transformations,
+the possible granularity of that configuration, and how to maximize observability
+of the effective resources requested by a Workload.
+
+### Specifying the Transformation
+
+A transformation matrix is defined as a single object.  Most likely
+this object would be embedded inside Kueue's `Configuration` API, but
+it would also be possible to make it a separate top-level config API or 
+use a config map containing a single yaml field that would be deserialized
+into the type defined below.  To support finer-grained configuration, instead
+of defining a single controller-wide transformation matrix, they could be embedded
+in ClusterQueue ResourceGroups (perhaps implicitly by instead embedding the
+transformation matrix in a ResourceFlavor).
+
+```go
+type ResourceTransformationStrategy string
+const Retain ResourceTransformationStrategy = "Retain"
+const Replace ResourceTransformationStrategy = "Replace"
+
+type ResourceTransformation struct {
+  // key is the output resource name and value is the quantity of the output resource 
+  // that corresponds to 1 unit of the input resource
+  ConversionBaseValues map[corev1.ResourceName]resource.Quantity `json:"conversionBaseValues"`
+
+  // should the input resource be replaced or retained
+  Strategy ResourceTransformationStrategy `json:"strategy"`
+}
+
+// The map key is the input resource name
+type ResourceTransformationMatrix map[corev1.ResourceName]ResourceTransformation
+
+```
+By using maps instead of arrays, we avoid needing
+to define the semantics of multiple entries having the same input or output
+resource name.  Using the input resource name as the key to the outer map
+avoids conflicting definitions of `Replace` for an input resource.
+
+As an example, the following list of 4-tuples
+
+* (`nvidia.com/mig-1g.5gb`, `example.com/accelerator-memory`, `5Gi`, `replace`)
+* (`nvidia.com/mig-2g.10gb`, `example.com/accelerator-memory`, `10Gi`, `replace`)
+* (`nvidia.com/mig-4g.20gb`, `example.com/accelerator-memory`, `20Gi`, `replace`)
+* (`nvidia.com/mig-1g.5gb`, `example.com/credits`, `10`, `replace`)
+* (`nvidia.com/mig-2g.10gb`, `example.com/credits`, `15`, `replace`)
+* (`nvidia.com/mig-4g.20gb`, `example.com/credits`, `30`, `replace`)
+* (`cpu`, `example.com/credits`, `1`, `retain`)
+
+would be represented in yaml when embedded in the Kueue Configuration object as
+```yaml
+resources:
+  excludedResourcePrefixes: []
+  transformations:
+    nvidia.com/mig-1g.5gb:
+      strategy: Replace
+      conversionBaseValues:
+        example.com/accelerator-memory: 5G
+        example.com/credits: 10
```

Comment:
What happens if both pre-transformed and post-transformed resource names are specified in the ClusterQueue in the following?
Will the kueue just treat the "nvidia.com/mig-1g.5gb" as the same as today? Or will we introduce any validation rules?

```yaml
# ClusterQueue
  - coveredResources: ["nvidia.com/mig-1g.5gb", "example.com/accelerator-memory"]
    flavors:
    - name: default-flavor
      resources:
      - name: "nvidia.com/mig-1g.5gb"
        nominalQuota: 10
      - name: example.com/accelerator-memory
        nominalQuota: 80G
---
# Configuration
resources:
  transformations:
    nvidia.com/mig-1g.5gb:
      strategy: Replace
      conversionBaseValues:
        example.com/accelerator-memory: 5G
```

Reply:
Only the post-transformed resources are used to assign flavors and reserve resources.  So the nominalQuota of `nvidia.com/mig-1g.5gb` given to the default-flavor by this cluster queue will never be requested.  In terms of the ClusterQueue logic, this is the same processing/logic that would happen when for example a workload that only requests `cpu` is admitted via using a flavor that covers both `cpu` and `memory`.

It is certainly would be possible to detect that a `coveredResource` is also an input resource with a `Replace` strategy.  I'm not sure what the appropriate level of warning/action is if this happens.  It seems like a parallel situation with a `coveredResource` being matching by an `excludedResourcePrefixes` which is not currently checked or validated by Kueue.

Reply:
Added discussion in KEP about interactions between `resourceMappings`,  `excludedResourcePrefixes` and `coveredResources`.